TESTING.md

```diff
@@ -0,0 +1,53 @@
+<!-- TOC -->
+- [Testing](#testing)
+    - [Code Coverage Collecting](#code-coverage-collecting)
+    - [Code Coverage Reporting](#code-coverage-reporting)
+<!-- TOC -->
+
+## Testing
+
+### Code Coverage Collecting
+Code coverage is a measurement of how many lines of code are executed under the test suites.
+
+#### OpenSearch plugin
+
+##### Gradle project
+1. Add [JaCoCo Gradle plugin](https://docs.gradle.org/current/userguide/jacoco_plugin.html) to `build.gradle` in your repository. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L48)
+2. Set Jacoco to export coverage report in `xml` format. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L101-L105)
+3. Call gradle task "jacocoTestReport" to generate report (`./gradlew jacocoTestReport`) after running your tests. (Or add a `dependsOn` to make "jacocoTestReport" task depends on your test tasks. [example](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle))
+
+##### Maven project
+Since OpenSearch uses Gradle as the build tool, we don't recommend using Maven.
+Code coverage for a Maven project can be generated by [JaCoCo Maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html), and we also got an [example](https://github.com/opensearch-project/security/blob/v1.0.0.0-rc1/pom.xml#L636).
+
+#### OpenSearch Dashboards plugin
+
+#### Limitation
+
+### Code Coverage Reporting
+OpenSearch plugins are using [Codecov](https://about.codecov.io/) for code coverage reporting and analysis. The dashboard can be seen [here](https://app.codecov.io/gh/opensearch-project/).
+
+All the OpenSearch plugins are required to have the following items:
+
+#### Code coverage report upload through CI workflow
+1. Generate code coverage report in `xml` format in a CI workflow of Github Actions
+2. Upload the report to Codecov by adding a step of [codecov-action](https://github.com/codecov/codecov-action) in the workflow.
+3. For private repositories, an upload token is needed, visit the repository page of Codecov to get your token: https://codecov.io/gh/opensearch-project/REPO-NAME-HERE. Please use Github Secrets to store the token and not expose to the public.
```

Comment:
Can we add a reference to Github secrets documentation and how it could be setup?

Reply:
I can add one, but I'm doubt if the plugins are required to use the codecov token. AD plugin is not using codecov token in your sample CI workflow, so I thought if that's fine, I don't need to take time in writing guide of Github Secrets. 😏

Reply:
Haha I didnt mean to add the whole example, just link github secrets documentation, folks should be able to figure out from that doc on how to add a secret to their repository  it if needed.
https://docs.github.com/en/actions/reference/encrypted-secrets

Reply:
Thanks for finding the link for me! I added that.